### PR TITLE
[IMP] pos: make receipt fields extensible

### PR DESCRIPTION
**Pull request by blaggacao**
Ticket Nr: 817016

Before this untangling, it was not possible to pass custom fields to the
to the ~receipt~ subreceipt xml expansion because it was processed within the constructor method.

Now the constructor method is carved out and can be cleanly inherited.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/point_of_sale/static/src/js/models.js [hunk 1]
```diff
@@ -1794,24 +1793,6 @@ exports.Order = Backbone.Model.extend({
...
-                subreceipt = subreceipt.split('\n').slice(1).join('\n');
-                var qweb = new QWeb2.Engine();
-                    qweb.debug = core.debug;
-                    qweb.default_dict = _.clone(QWeb.default_dict);
-                    qweb.add_template('<templates><t t-name="subreceipt">'+subreceipt+'</t></templates>');
-                
-                return qweb.render('subreceipt',{'pos':self.pos,'widget':self.pos.chrome,'order':self, 'receipt': receipt}) ;
-            }
-        }
-
```
Comment (blaggacao) [inline addons/point_of_sale/static/src/js/models.js]: Only this...

File: addons/point_of_sale/static/src/js/models.js [hunk 2]
```diff
@@ -1857,7 +1838,30 @@ exports.Order = Backbone.Model.extend({
...
+                var qweb = new QWeb2.Engine();
+                    qweb.debug = core.debug;
+                    qweb.default_dict = _.clone(QWeb.default_dict);
+                    qweb.add_template('<templates><t t-name="subreceipt">'+subreceipt+'</t></templates>');
+
+                return qweb.render('subreceipt',{'pos':self.pos,'widget':self.pos.chrome,'order':self, 'receipt': receipt}) ;
+            }
+        }
+
         if (is_xml(this.pos.config.receipt_header)){
```
Comment (blaggacao) [inline addons/point_of_sale/static/src/js/models.js]: And this...

File: addons/point_of_sale/static/src/js/models.js [hunk 3]
```diff
@@ -1857,7 +1838,30 @@ exports.Order = Backbone.Model.extend({
...
+                subreceipt = subreceipt.split('\n').slice(1).join('\n');
+                var qweb = new QWeb2.Engine();
+                    qweb.debug = core.debug;
+                    qweb.default_dict = _.clone(QWeb.default_dict);
+                    qweb.add_template('<templates><t t-name="subreceipt">'+subreceipt+'</t></templates>');
+
+                return qweb.render('subreceipt',{'pos':self.pos,'widget':self.pos.chrome,'order':self, 'receipt': receipt}) ;
+            }
+        }
+
```
Comment (blaggacao) [inline addons/point_of_sale/static/src/js/models.js]: And this...

**Comment by blaggacao** [comment]
@jorenvo receipt WAS constructed using a variable attribute on the "model". I found no decent way to alter this variable, so I made it to be passed as an (inheritable) function argument.

**Comment by jorenvo** [comment]
I don't get why this is necessary. You can pass custom fields to the receipt. See: https://github.com/odoo/odoo/blob/3285a0384c1bdabc5d6c9c906a7ed54d3e595e7c/addons/pos_restaurant/static/src/js/floors.js#L758-L764

**Comment by blaggacao** [comment]
@jorenvo Thanks for your reply. The reason is the rendering of the subreceipt (sic!) is done in the render_xml and the result is the exported for printing. Therefore there are two rendering steps:

1. subreceipt
2. Whole receipt based on the json (at this point the subrecipt is already rendered.

see:
https://github.com/odoo/odoo/blob/11.0/addons/point_of_sale/static/src/js/models.js#L2075
```javascript
        function render_xml(subreceipt){
            if (!is_xml(subreceipt)) {
                return subreceipt;
            } else {
                subreceipt = subreceipt.split('\n').slice(1).join('\n');
                var qweb = new QWeb2.Engine();
                    qweb.debug = core.debug;
                    qweb.default_dict = _.clone(QWeb.default_dict);
                    qweb.add_template('<templates><t t-name="subreceipt">'+subreceipt+'</t></templates>');

                return qweb.render('subreceipt',{'pos':self.pos,'widget':self.pos.chrome,'order':self, 'receipt': receipt}) ;
            }
```

**Comment by blaggacao** [comment]
Sorry, I see my commit message wasn't clear on that is about subreceipts...

**Comment by blaggacao** [comment]
@jorenvo Could you consider to merge this, a whole branch of custom developments depends on this by now...

**Comment by blaggacao** [comment]
@odony help as well, just an enabler and I think it's sufficiently documented.

**Comment by Yenthe666** [comment]
Probably something for @mart-e as (@)odony is more for framework & security, not for POS.
@mart-e can you give it some TLC please?

**Comment by jorenvo** [comment]
Let my try to clarify what I said before, because I still don't think you need this. Receipts are rendered by the `ReceiptScreenWidget` using the same rendering context for both the HTML receipt (`PosTicket` and the receipt printer receipt (`XmlReceipt`). The main part of this rendering context comes from the `export_for_printing` functions on the `Order`, `Orderline`, and `Paymentline` models.

You seem to be interested in modifying the behavior of the 'subreceipt' stuff. This is a fairly unknown and undocumented feature introduced a long time ago and it should not be used in your case. This feature was introduced so that on our SaaS platform users could insert custom QWeb XML in the `receipt_header` and `receipt_footer` fields on `pos.config` that would then be rendered. Since the receipts are Javascript QWeb templates they cannot easily be customized like other Odoo views (i.e. they are not `ir.ui.view` records). I suppose at the time this was considered an acceptable hack.

In your case you should not be relying on this feature. Since you are developing custom modules you don't need this hack. You should instead modify both `PosTicket` and `XmlReceipt` using [Javascript QWeb template inheritance](https://www.odoo.com/documentation/11.0/reference/qweb.html#template-inheritance). This is done in various places in standard Odoo, e.g. the pos_restaurant module adds the table name and the amount of guests to the top of the receipt as follows:

https://github.com/odoo/odoo/blob/e449055189328d999a1d61a778dca0aabd34426b/addons/pos_restaurant/static/src/xml/floors.xml#L4-L13

For this to work correctly the `export_for_printing` function was modified to include this extra information (like I linked before):

https://github.com/odoo/odoo/blob/3285a0384c1bdabc5d6c9c906a7ed54d3e595e7c/addons/pos_restaurant/static/src/js/floors.js#L758-L764

So in summary I would recommend migrating your developments to define new templates using `t-extend` and modifying the render context like above. This way you don't have to rely on this hacky subreceipt rendering (with a render context that is admittedly hard to modify ;)).

If you still disagree, could you give a small but complete code example of what you're doing now?

**Comment by blaggacao** [comment]
You are right in my special use case I cannot think of a valid customer requirement where the customer might have sustainable argument to modify the appearance of this:
```xml
<div font="b">----------------------------------------------------</div>
<br />
<t t-if="order.sequence_number_uniq"><b>Factura de Venta No.: <t t-esc="order.pos.pos_session.prefix" /><t t-esc="order.sequence_number_uniq" /></b></t>
<t t-if="!order.sequence_number_uniq"><b>Recibo de Caja</b></t>
<br />
<div font="b">----------------------------------------------------
<br />
Cliente: <t t-esc="receipt.client.name || '  ........................................' " />
<br />
Raz.Soz: <t t-esc="receipt.client.legalname || '  ........................................' " />
<br />
IdFisca: <t t-esc="receipt.client.co_nit || '  ........................................' " />
<br />
----------------------------------------------------</div>
<br />
```

I think migrating the code is acceptable if this is an undocumented (unstable) feature.

**Comment by AbrahamMatu** [comment]
Dear Gurus,

Could you assist in making table numbers appear on POS restaurant receipt?

Abraham

**Comment by Yenthe666** [comment]
@AbrahamMatu please do ask these questions on the official help forum at help.odoo.com
